File: apps/keyboard-android/app/src/main/java/com/intelliai/keyboard/audio/WavEncoder.kt
package com.intelliai.keyboard.audio

import java.io.ByteArrayOutputStream

/**
 * PCM16 → WAV container, as a pure function — no Android types, fully
 * unit-tested. The backend's canonical pipeline resamples everything
 * server-side, so the keyboard ships exactly what it recorded: 16 kHz,
 * mono, 16-bit little-endian PCM in a standard RIFF/WAVE/fmt/data
 * layout. No client-side resampling, no encoders, no audio libraries.
 */
object WavEncoder {

    const val SAMPLE_RATE_HZ = 16_000
    const val CHANNELS = 1
    const val BITS_PER_SAMPLE = 16

    fun wrapPcm16(
        pcm: ByteArray,
        sampleRateHz: Int = SAMPLE_RATE_HZ,
        channels: Int = CHANNELS,
    ): ByteArray {
        val bytesPerSample = BITS_PER_SAMPLE / 8
        val byteRate = sampleRateHz * channels * bytesPerSample
        val blockAlign = channels * bytesPerSample
        val out = ByteArrayOutputStream(44 + pcm.size)

        out.writeAscii("RIFF")
        out.writeIntLe(36 + pcm.size) // RIFF chunk size: rest of file
        out.writeAscii("WAVE")

        out.writeAscii("fmt ")
        out.writeIntLe(16) // PCM fmt chunk length
        out.writeShortLe(1) // audio format 1 = linear PCM
        out.writeShortLe(channels)
        out.writeIntLe(sampleRateHz)
        out.writeIntLe(byteRate)
        out.writeShortLe(blockAlign)
        out.writeShortLe(BITS_PER_SAMPLE)

        out.writeAscii("data")
        out.writeIntLe(pcm.size)
        out.write(pcm)

        return out.toByteArray()
    }

    /** Duration the PCM payload represents, for limits and UX. */
    fun durationMs(pcmByteCount: Int, sampleRateHz: Int = SAMPLE_RATE_HZ): Long {
        val bytesPerSecond = sampleRateHz * CHANNELS * (BITS_PER_SAMPLE / 8)
        return pcmByteCount * 1000L / bytesPerSecond
    }

    private fun ByteArrayOutputStream.writeAscii(text: String) =
        write(text.toByteArray(Charsets.US_ASCII))

    private fun ByteArrayOutputStream.writeIntLe(value: Int) {
        write(value and 0xFF)
        write(value ushr 8 and 0xFF)
        write(value ushr 16 and 0xFF)
        write(value ushr 24 and 0xFF)
    }

    private fun ByteArrayOutputStream.writeShortLe(value: Int) {
        write(value and 0xFF)
        write(value ushr 8 and 0xFF)
    }
}

File: apps/keyboard-ios/Shared/WavEncoder.swift
import Foundation

/// PCM16 → WAV container, as a pure function — no Apple audio types,
/// fully unit-testable. Canonicalization happens server-side, so the
/// keyboard ships exactly what it recorded: 16 kHz, mono, 16-bit
/// little-endian PCM in a standard RIFF/WAVE/fmt/data layout —
/// byte-compatible with the Android client's encoder and the backend's
/// canonical audio path.
enum WavEncoder {
    static let sampleRateHz: Int = 16_000
    static let channels: Int = 1
    static let bitsPerSample: Int = 16

    /// Wrap little-endian PCM16 sample bytes in a RIFF/WAVE header.
    static func wav(fromPCM pcm: Data) -> Data {
        let byteRate = sampleRateHz * channels * bitsPerSample / 8
        let blockAlign = channels * bitsPerSample / 8

        var data = Data(capacity: 44 + pcm.count)
        data.append(contentsOf: Array("RIFF".utf8))
        data.appendUInt32LE(UInt32(36 + pcm.count))
        data.append(contentsOf: Array("WAVE".utf8))
        data.append(contentsOf: Array("fmt ".utf8))
        data.appendUInt32LE(16) // fmt chunk size
        data.appendUInt16LE(1) // PCM
        data.appendUInt16LE(UInt16(channels))
        data.appendUInt32LE(UInt32(sampleRateHz))
        data.appendUInt32LE(UInt32(byteRate))
        data.appendUInt16LE(UInt16(blockAlign))
        data.appendUInt16LE(UInt16(bitsPerSample))
        data.append(contentsOf: Array("data".utf8))
        data.appendUInt32LE(UInt32(pcm.count))
        data.append(pcm)
        return data
    }
}

private extension Data {
    mutating func appendUInt32LE(_ value: UInt32) {
        var little = value.littleEndian
        Swift.withUnsafeBytes(of: &little) { append(contentsOf: $0) }
    }

    mutating func appendUInt16LE(_ value: UInt16) {
        var little = value.littleEndian
        Swift.withUnsafeBytes(of: &little) { append(contentsOf: $0) }
    }
}
